File: 2024_12/leetcode2024-12-06.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int numRookCaptures(vector<vector<char>>& board) {
        int x=0, y=0;
        for (int i=0; i<8; i++){
            for (int j=0; j<8; j++){
                if (board[i][j] == 'R'){
                    x=i;
                    y=j;
                    break;
                }
            }
        }
        vector<vector<int>> directions = {{0,-1},{0,1},{1,0},{-1,0}};
        int res = 0;
        for (auto d : directions){
            int dx = d[0], dy = d[1];
            int nx = x+dx, ny = y+dy;
            while ((nx>=0 && nx<8) && (ny>=0 && ny<8)){
                if (board[nx][ny] == 'B'){
                    break;
                }
                if (board[nx][ny] == 'p'){
                    res++;
                    break;
                }
                nx += dx;
                ny += dy;
            }
        }
        return res;
    }
};

int main(){
    Solution s;
    vector<vector<char>> board = {{'.','.','.','.','.','.','.','.'},{'.','.','.','p','.','.','.','.'},{'.','.','.','R','.','.','.','p'},{'.','.','.','.','.','.','.','.'},{'.','.','.','.','.','.','.','.'},{'.','.','.','p','.','.','.','.'},{'.','.','.','.','.','.','.','.'},{'.','.','.','.','.','.','.','.'}};
    cout << s.numRookCaptures(board) << endl;
}

File: leetcode2024/2024_12/leetcode2024-12-06.py
from typing import List

class Solution:
    def numRookCaptures(self, board: List[List[str]]) -> int:
        x, y = 0, 0
        for i in range(8):
            for j in range(8):
                if board[i][j] == 'R':
                    x, y = i, j
                    break
        directions = [[0,-1],[0,1],[1,0],[-1,0]]
        res = 0
        for dx,dy in directions:
            nx,ny = x+dx,y+dy
            while (nx>=0 and nx<8) and (ny>=0 and ny<8):
                if board[nx][ny] == 'B':
                    break
                if board[nx][ny] == 'p':
                    res += 1
                    break
                nx,ny = nx+dx,ny+dy
        return res

if __name__ == "__main__":
    s = Solution()
    board = [[".",".",".",".",".",".",".","."],[".",".",".","p",".",".",".","."],[".",".",".","R",".",".",".","p"],[".",".",".",".",".",".",".","."],[".",".",".",".",".",".",".","."],[".",".",".","p",".",".",".","."],[".",".",".",".",".",".",".","."],[".",".",".",".",".",".",".","."]]
    print(s.numRookCaptures(board))
